### ISSUE TITLE
Request: Make number entry to stop at word size limit instead of overwriting MSB.

### ISSUE THREAD

**Issue by meltdown03** [zathras/jrpn#53]
**Is your feature request related to a problem? Please describe.**
I don't know if the real 16C does this or not, but it may affect the realism.

**Describe the solution you'd like**
Add an option to ***not*** allow more numbers to be entered once you hit the Word Size of the current mode.

**Describe any alternatives you've considered**
I tried looking at the source code but I don't have much experience with Android apps.

**Additional context**
I have the Repo synced and building so if there is a quick change in the code, I can make it locally for my needs.

**Comment by zathras** [comment]
Ah, no, I'm confident that this is the behavior of a real 16C.  Stopping at the word size would have, in some ways, be easier; it took some doing to accept the entry modulo the current word size, then display it in the current mode.  I also checked this behavior with other simulators.

See the manual starting on page 37 ("The Display and Internal Representation"), where entering "254" in 8 bit signed decimal 1's complement mode results in "-001" on the display.  (The manual may still be under copyright, so I don't link to it, but it's easy to find.)

As to making settings that change the behavior away from what the 16C does, I generally shy away from that (except for "disable window," because seeing a 64 bit binary number in magically shrinking LCD digits was easy and fun).  As a general rule, my theory is that better calculators were made after the early '80s anyway.

**Comment by meltdown03** [comment]
Is this the section you are talking about?? Just curious. I was using the WRPN 16c on android and the HP16c for windows and they both stop once you get to the word size.
``` java
@override
String format(Value v, Model m) {
  String s = v.internal.toRadixString(radix);
  if (_leadingZeroesOK && m.displayLeadingZeros) {
    final int digits = (m.wordSize + _bitsPerDigit - 1) ~/ _bitsPerDigit;
    final int n = digits - s.length;
    // 64 zeroes (63 would actually do):
    s = '0000000000000000000000000000000000000000000000000000000000000000'
            .substring(0, n) +
        s;
  }
  return addCommas(s, m.settings.integerModeCommas) + displayName;
}
```

**Comment by meltdown03** [comment]
I also found that the Legacy java version beeps when you try to hit a fifth digit in 16 bit word size. It doesn't pop off the most significant digit like this version.

**Comment by meltdown03** [comment]
Here is a video of what I see, it's very strange in DEC mode: 
[Screencast from 2023-07-23 20-04-02.webm](https://github.com/zathras/jrpn/assets/6166536/9c11a915-7fa6-4a70-a7e1-e77797a388d7)

**Comment by zathras** [comment]
That looks like code involved in this, but I was just talking about the behavior as described in the HP 16c Owner's Manual, page 37.

"JRPN Legacy" was basically WRPN with a different UI; in JRPN Legacy I didn't look at changing the calculator behavior from WRPN.  With the new JRPN I wrote it from scratch, paying very close attention to the Owner's Manual.

**Comment by meltdown03** [comment]
I see what you mean now, thanks for the clarification. I ran through the steps on both sims and yours did what the manual said, the other sim needed me to hit Enter to show -1

**Comment by zathras** [comment]
Yeah, when I was going though the Owner's Handbook page by page, in excruciating detail, I remember reading that example on page 37.  My initial reaction was, "wait, what?"  It took me a while to figure out that they were just blindly doing the operation implied by a digit entry, throwing away the higher order bits, and then displaying whatever that came out as in the current display mode.  I bet that worked out as easier to implement on a 4 bit processor in the early '80s (or maybe late '70s when they wrote it), and it makes a sense from a certain perspective, but it's really weird to see that behavior by modern UI standards.

To me, figuring out these behaviors and sticking to what the original did was part of the fun.  Another one in that vein is what happens to the registers when you change the word size, or switch between integer and floating point mode.  The bits in register memory don't change, but their interpretation does.  Fun stuff!

**Comment by meltdown03** [comment]
I did find out from this video that it doesn't let you keep pressing buttons: https://youtu.be/bRvGoRbR_bA at time 5:35. 
Not sure if there's a way to keep all the other things you have figured out.

**Comment by zathras** [comment]
Oh, that's interesting, thanks.  Maybe it's allowing an unsigned->signed conversion, but not allowing something that overflows?

If that's right, I guess the behavior in 8 bit decimal signed mode would be:
255777777 give -1
254777777 gives -2
256777777 gives 25

For a more challenging example, 9 bit decimal signed mode would be:
2567777777 gives -256
1023777777 gives gives -1
1024777777 gives 102

I'll need to chase down someone with a real 16C to see if this guess is right...

**Comment by meltdown03** [comment]
I'm bidding on one now on eBay but I'll probably get sniped.

**Comment by zathras** [comment]
Posted https://www.hpmuseum.org/forum/thread-20239.html with two extra examples.

**Comment by meltdown03** [comment]
Would the go16c android app work to show the actual behavior since it uses
ROM dumps? I bought it but then refunded it once already, lol.

**Comment by zathras** [comment]
> Would the go16c android app work to show the actual behavior since it uses ROM dumps? I bought it but then refunded it once already, lol.

I'm not familiar with it, but probably.  However, folks on the HP Museum forum are generally quite responsive about this kind of thing.  Since this isn't a commercial product or anything, like Blanche Dubois, I'm more than happy to depend on the kindness of strangers.

**Comment by meltdown03** [comment]
The emulator from http://www.hp16c.net does what you assumed but unfortunately no source code

**Comment by zathras** [comment]
For completeness, the Forum verified the following behavior.  8 bit decimal signed 2's complement mode when pressing digits without pressing enter:
255777777 gives -1
254777777 gives -2
256777777 gives 25
256477777 gives -2 (the 6 is ignored, but the 4 isn't)

and in 10 bit decimal signed 2's complement mode:
5127777777 gives -512
1023777777 gives gives -1
1024777777 gives 102
1024177777 gives -3

> Confirmed on my 16C (2251A) though the keys can sometimes not register. g-Clx between entries. ... no overflow flag. 

The code to do this should be pretty siimple.

**Comment by meltdown03** [comment]
Do you think you should ask what happens when you hit BSP on those tests?
Also, I didn't win that auction, partly because I wasn't paying attention at the end of it. I bid on a different auction that ends in 3 days, but it's for 2 16C's and a 12C so it might get too pricey.

**Comment by zathras** [comment]
> Do you think you should ask what happens when you hit BSP on those tests?

Oh, good idea.  I wouldn't be surprised if BSP turned into CLX in some of these cases.

**Comment by zathras** [comment]
It looks like it keeps a flag of whether the currently displayed number came from an unsigned->signed conversion, and on BSP, if the flag is set, it reverses the conversion.  8 bit signed decimal 2's complement:
```
CLX     //    0 d
2       //    2 d
5       //   25 d
5       // -001 d
BSP     //   25 d
```
Not what I was expecting, but perfectly reasonable, and downright clever.

Also, just making sure about overflow, in 8 bit signed decimal:
```
CLx       0 d
258       25 d
4         254 d
BSP       25 d
BSP       2 d
```
Credit:  https://www.hpmuseum.org/forum/thread-20239-post-175219.html#pid175219

**Comment by zathras** [comment]
Hi @meltdown03,

I think this latest putback should implement the new, more-faithful behavior.  Could you build this and see if it behaves as expected?

The algorithm really is pretty simple, but by the time it's integrated into the input state machine, which is designed to work in various modes, the code becomes a little subtle and complex.  But I think I covered all the bases!

**Comment by meltdown03** [comment]
Yes, this seems to work how I expected now. Very nice!

**Comment by zathras** [comment]
Released in 2.1.7
